src/main/java/org/jitsi/jibri/api/xmpp/XmppApi.kt

```diff
@@ -0,0 +1,214 @@
+import org.jitsi.jibri.util.getCallUrlInfoFromJid
+import org.jitsi.xmpp.mucclient.MucClient
+import org.jivesoftware.smack.packet.IQ
+import org.jivesoftware.smack.packet.XMPPError
+import org.jivesoftware.smack.provider.ProviderManager
+import org.jivesoftware.smack.tcp.XMPPTCPConnectionConfiguration
+import org.jxmpp.jid.impl.JidCreate
+import org.jxmpp.jid.parts.Resourcepart
+import java.util.concurrent.Executors
+import java.util.logging.Logger
+
+/**
+ * [XmppApi] connects to XMPP MUCs according to the given [xmppConfigs] (which are parsed from config.json)
+ * and listens for IQ messages which contain Jibri commands, which it relays
+ * to the given [JibriManager].
+ * The IQ messages are instances of [JibriIq] and allow the starting and stopping of
+ * the services Jibri provides.
+ * [XmppApi] subscribes to [JibriManager] status updates and translates those into
+ * XMPP presence (defined by [JibriStatusPacketExt]) updates to advertise the status of this Jibri.
+ * [XmppApi] takes care of translating the XMPP commands into the appropriate
+ * [JibriManager] API calls and translates the results into XMPP IQ responses.
+ */
+class XmppApi(
+    private val jibriManager: JibriManager,
+    private val xmppConfigs: List<XmppEnvironmentConfig>
+) {
+    private val logger = Logger.getLogger(this::class.qualifiedName)
+    private val executor = Executors.newSingleThreadExecutor(NameableThreadFactory("XmppApi"))
+
+    /**
+     * Start up the XMPP API by connecting and logging in to all the configured XMPP environments.  For each XMPP
+     * connection, we'll listen for incoming [JibriIq] messages and handle them appropriately.  Join the MUC on
+     * each connection and send an initial [JibriStatusPacketExt] presence.
+     */
+    fun start() {
+        JibriStatusPacketExt.registerExtensionProvider()
+        ProviderManager.addIQProvider(
+            JibriIq.ELEMENT_NAME, JibriIq.NAMESPACE, JibriIqProvider()
+        )
+
+        // Join all the mucs we've been told to
+        for (config in xmppConfigs) {
+            for (host in config.xmppServerHosts) {
+                logger.info("Connecting to xmpp environment on $host with config $config")
+                val mucClient = MucClient(XMPPTCPConnectionConfiguration.builder()
+                    .setHost(host)
+                    .setXmppDomain(config.controlLogin.domain)
+                    .setUsernameAndPassword(config.controlLogin.username, config.controlLogin.password)
+                    .build())
+                mucClient.addIqRequestHandler(object : JibriSyncIqRequestHandler() {
+                    override fun handleJibriIqRequest(jibriIq: JibriIq): IQ {
+                        return handleJibriIq(jibriIq, config, mucClient)
+                    }
+                })
+                jibriManager.addStatusHandler { status ->
+                    logger.info("XMPP API got jibri status $status, publishing presence to connection ${config.name}")
```

Comment:
Is there any config validation? Is that even needed?

Reply:
config is validated when it's parsed (in terms of the proper fields being present).  there may be some more semantic validating we could do on the field values themselves...maybe worth doing some negative testing to make sure things are handled ok.